debian/drush.1
.\"                                      Hey, EMACS: -*- nroff -*-
'\" t -*- coding: UTF-8 -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH DRUSH 1 "January 11, 2011"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
drush \- shell scripting interface for Drupal
.SH SYNOPSIS
.B drush
.RI [ options ] " command" ...
.SH DESCRIPTION
.PP
\fBdrush\fP is a veritable Swiss Army knife designed to make life
easier for those of us who spend some of our working hours hacking
away at the command prompt.
.PP
\fBDrush\fP core ships with lots of useful commands for interacting with
code like modules/themes/profiles. Similarly, it runs update.php,
executes sql queries and DB migrations, and misc utilities like run
cron or clear cache.
.SH OPTIONS
.TP
.BI \-r\  path ,\ \-\-root= path
Drupal root directory to use (default: current directory)             
.TP
.BI \-l\  uri ,\ \-\-uri= uri
URI of the drupal site to use (only needed in multisite environments) 
.TP
.BI \-v ,\ \-\-verbose
Display extra information about the command.                          
.TP
.B \-q, \-\-quiet
Hide all output
.TP
.B \-d, \-\-debug
Display even more information, including internal messages.           
.TP
.B \-y, \-\-yes
Assume 'yes' as answer to all prompts                                 
.TP
.B \-s, \-\-simulate
Simulate all relevant actions (don't actually change the system)
.TP
.B \-p, \-\-pipe
Emit a compact representation of the command for scripting.           
.TP
.B \-h, \-\-help
A detailed help system, use \fBdrush help help\fP for complete help.
.TP
.B \-\-version
Show drush version.
.PP
If you get tired of typing options all the time, you can add them to
your drush.php alias or create a drushrc.php file. These provide
additional options for your drush call. They provide great flexibility
for a multi-site installation, for example. See example.drushrc.php.
.SH COMMANDS
Drush is all based around "commands" that are usually defined in
various modules. 
.TP
.B help
Print the help message. Use \-\-filter to limit command list to one
command file (e.g. \-\-filter=pm).
.TP
.B cache-clear (cc)
Clear a specific cache, or all drupal caches.
.TP
.B core-cli (cli)
Enter a new shell optimized for Drush use.
.TP
.B core-cron (cron)
Run all cron hooks.
.TP
.B core-status (st)
Provides a birds-eye view of the current Drupal installation, if any.
.TP
.B php-eval (eval)
Evaluate arbitrary php code after bootstrapping Drupal.
.TP
.B updatedb (updb)
Apply any database updates required (as with running update.php).
.TP
.B variable-get (vget)
Get a list of some or all site variables and values
.TP
.B variable-set (vset)
Set a variable.
.TP
.B watchdog show
Shows recent watchdog log messages. Optionally filter for a specific type.
.TP
.B pm-enable (en)
Enable one or more extensions (modules or themes).
.TP
.B pm-disable (dis)
Disable one or more extensions (modules or themes).
.TP
.B pm-download (dl)
Download projects from drupal.org or other sources.
.TP
.B pm-uninstall
Uninstall one or more modules.
.TP
.B pm-updatecode (upc)
Update Drupal core and contrib projects to latest recommended releases.
.TP
.B pm-update (up)
Update Drupal core and contrib projects and apply any pending database
updates (Same as pm-updatecode + updatedb).
.TP
.B sql-cli (sqlc)
Open a SQL command-line interface using Drupal's credentials.
.TP
.B sql-connect
A string for connecting to the DB.
.TP
.B sql-dump
Exports the Drupal DB as SQL using mysqldump.
.PP
This manual only lists a few of the most frequently used builtin
commands shipped with the Drush core, since other Drupal modules and
third-party "command files" may extend this list significantly. The
canonical source of information for the available commands and their
usage is the online help system.
.PP
More information on each command is also available through \fBdrush
help <command>\fP. Most commands have shortcuts indicated in
parenthesis.
.SH FILES
.IR /etc/drush/drushrc.php
.br
.I /usr/share/doc/drush/examples/example.drushrc.php
.SH SEE ALSO
.BR cvs (1),
.BR svn (1),
.BR wget (1).
.SH AUTHOR
Drush was originally developed by Arto for Drupal 4.7 (this alpha code
can still be found in the DRUPAL-4-7 branch). In May 2007, it was
partly rewritten and redesigned for Drupal 5 by frando. Since 2008,
the module has been maintained by Moshe Weitzman, Owen Barton and
Adrian Rossouw.
.PP
This manual page was written by Antoine Beaupré <anarcat@koumbit.org>
in June 2009 for the Debian project (but may be used by others).


debian/changelog
drush (4.5-1) unstable; urgency=low

  * new upstream release: http://drupal.org/node/1247086
  * fix watch file to follow new naming convention
  * fix a bunch of lintian tips
  * switch to source format 3.0 (quilt)
  
 -- Antoine Beaupré <anarcat@koumbit.org>  Sun, 21 Aug 2011 15:14:15 +0100

drush (4.4-2) unstable; urgency=low

  * ship properly all the examples
  * ship properly the docs/ directory for drush topic to work
  * update to standards version 3.9.2, no change
  * do not suggest subversion/cvs but git (Closes: #628794).

 -- Antoine Beaupré <anarcat@koumbit.org>  Tue, 28 Jun 2011 14:15:06 -0400

drush (4.4-1) unstable; urgency=low

  * New upstream minor release to fix .info file.

 -- Antoine Beaupré <anarcat@koumbit.org>  Mon, 14 Mar 2011 19:22:40 -0400

drush (4.3-1) unstable; urgency=low

  * New upstream release, see http://drupal.org/node/1091572
  * Changed git repos to be the upstream repos, which have accepted that
    the Debian branch live upstream. Upstream (msonnabaum) has also
    manifested interest in maintaining the Debian/Ubuntu packages.

 -- Antoine Beaupré <anarcat@koumbit.org>  Mon, 14 Mar 2011 17:11:53 -0400

drush (4.2-1) unstable; urgency=low

  * New upstream release
  * Updated manpage (Closes: #595908)
  * Forgot to mention that the upstream release fixed the vget bug
    (Closes: #595772)

 -- Antoine Beaupré <anarcat@koumbit.org>  Thu, 20 Jan 2011 22:34:57 -0500

drush (4.1-1) unstable; urgency=low

  * New upstream release (Closes: #609362)
    This branch (4.x) is likely to live only in sid considering upstream
    is already considering a 5.x release. In general, the release cycle of
    Debian and Drush are incompatible, the latter being roughly every 6
    months. I will try to keep sid (and therefore testing) up to date with
    the upstream stable and people will be free to use 3.3 (in squeeze or
    lenny-backports) or whatever the latest stable upstream is (in sid or
    testing).

  * standards upgrade to 3.9.1 - no change
  * update table.inc to follow upstream

 -- Antoine Beaupré <anarcat@koumbit.org>  Thu, 20 Jan 2011 21:06:50 -0500

drush (3.3-1) unstable; urgency=low

  * New upstream version (Closes: #594020)
    This is the last "unclean" version of Drush 3. Further version will
    contain only security upgrades and critical bugfixes that can be
    targeted at Debian stable. As the maintainer of the Drush stable
    branch, I can commit to such a policy.

 -- Antoine Beaupré <anarcat@koumbit.org>  Fri, 27 Aug 2010 20:43:25 +0200

drush (3.1-2) unstable; urgency=low

  * fix version control urls
  * fix the drush.info issue (Closes: #588639)

 -- Antoine Beaupré <anarcat@koumbit.org>  Sun, 22 Aug 2010 19:20:02 +0200

drush (3.1-1) unstable; urgency=low

  * new upstream release, see http://drupal.org/node/833938
  * last upload mistakenly included a .diff.gz against head, this upload
    should correct that.

 -- Antoine Beaupré <anarcat@koumbit.org>  Wed, 23 Jun 2010 20:13:58 -0400

drush (3.0-1) unstable; urgency=low

  * new upstream release, see http://drupal.org/node/779448 
    (Closes: #575880).
  * removed all local patches integrated upstream
  * update to standards 3.8.4, no change required

 -- Antoine Beaupré <anarcat@koumbit.org>  Wed, 23 Jun 2010 19:53:13 -0400

drush (2.0-6) unstable; urgency=low

  * put this back to GPL2+ after discussions with Drupal.org people
  * update to standards 3.8.3: no change required
  * fix subversion dependencies, push php depends up to Recommends

 -- Antoine Beaupré <anarcat@koumbit.org>  Fri, 11 Sep 2009 22:47:11 +0200

drush (2.0-5) unstable; urgency=low

  * fix copyright notice to claim GPL-2 (and not GPL-2+)
  * update to standards 3.8.2: put in php section
  * backport a few fixes that have been committed to CVS after the 2.0
    release
    - #500324 dman - fails to download install profile in proper
      location. Incorrect path.
    - #493456 by guidot. three tiny errors in drush help dl Examples
      ... Also a text improvement.
    - #503328 - fix drush cron due to drush shutdown exiting. register
      exit helper in drush_shutdown. Thanks david goode
    - #533496 ZoFreX - Drush does not like being invoked from a path that
      contains spaces

 -- Antoine Beaupré <anarcat@koumbit.org>  Wed, 29 Jul 2009 17:52:18 -0400

drush (2.0-5~koumbit50+1) stable; urgency=low

  * backport to lenny

 -- Antoine Beaupré <anarcat@koumbit.org>  Wed, 29 Jul 2009 17:48:18 -0400

drush (2.0-4) unstable; urgency=low

  * fix paths in VCS fields
  * add DM-Upload-Allowed field since I am a Debian Maintainer
  * add VCS-* fields to debian/control file

 -- Antoine Beaupré <anarcat@koumbit.org>  Tue, 14 Jul 2009 21:21:00 -0400

drush (2.0-4~koumbit50+1) stable; urgency=low

  * backport to lenny
  * add VCS-* fields to debian/control file

 -- Antoine Beaupré <anarcat@koumbit.org>  Wed, 17 Jun 2009 21:07:18 -0400

drush (2.0-3) unstable; urgency=low

  * use dh_prep instead of dh_clean to silence warnings on sid. this will
    now fail to build on lenny.
  * cleanup comments in watch and rules files
  * update to standards 3.8.1

 -- Antoine Beaupré <anarcat@koumbit.org>  Sun, 14 Jun 2009 17:10:59 -0400

drush (2.0-2) unstable; urgency=medium

  * don't use the bash wrapper because it breaks in cron

 -- Antoine Beaupré <anarcat@koumbit.org>  Sun, 14 Jun 2009 16:47:43 -0400

drush (2.0-1) unstable; urgency=low

  * Initial release (Closes: #532723)
  * Write a manpage
  * Includes the table.inc file from PEAR's Console_Table

 -- Antoine Beaupré <anarcat@koumbit.org>  Thu, 11 Jun 2009 21:56:06 -0400



drush.php
#!/usr/bin/env php
<?php

/**
 * @file
 * drush is a PHP script implementing a command line shell for Drupal.
 *
 * @requires PHP CLI 5.2.0, or newer.
 */

require(dirname(__FILE__) . '/includes/bootstrap.inc');

drush_bootstrap_prepare();
exit(drush_main());

/**
 * The main Drush function.
 *
 * - Runs "early" option code, if set (see global options).
 * - Parses the command line arguments, configuration files and environment.
 * - Prepares and executes a Drupal bootstrap, if possible,
 * - Dispatches the given command.
 *
 * function_exists('drush_main') may be used by modules to detect whether
 * they are being called from drush.  See http://drupal.org/node/1181308
 * and http://drupal.org/node/827478
 *
 * @return
 *   Whatever the given command returns.
 */
function drush_main() {
  $return = '';
  if ($file = drush_get_option('early', FALSE)) {
    require_once($file);
    $function = 'drush_early_' . basename($file, '.inc');
    if (function_exists($function)) {
      if ($return = $function()) {
        // If the function returns FALSE, we continue and attempt to bootstrap
        // as normal. Otherwise, we exit early with the returned output.
        if ($return === TRUE) {
          $return = '';
        }
        drush_bootstrap_finish();
        return $return;
      }
    }
  }

  // Process initial global options such as --debug.
  _drush_bootstrap_global_options();

  $phases = _drush_bootstrap_phases(FALSE, TRUE);

  $return = '';
  $command_found = FALSE;

  foreach ($phases as $phase) {
    if (drush_bootstrap_to_phase($phase)) {
      // If applicable swaps in shell alias value (or executes it).
      drush_shell_alias_replace();

      $command = drush_parse_command();

      // Process a remote command if 'remote-host' option is set.
      if (drush_remote_command()) {
        $command_found = TRUE;
        break;
      }

      if (is_array($command)) {
        $bootstrap_result = drush_bootstrap_to_phase($command['bootstrap']);
        drush_enforce_requirement_bootstrap_phase($command);
        drush_enforce_requirement_core($command);
        drush_enforce_requirement_drupal_dependencies($command);
        drush_enforce_requirement_drush_dependencies($command);

        if ($bootstrap_result && empty($command['bootstrap_errors'])) {
          drush_log(dt("Found command: !command (commandfile=!commandfile)", array('!command' => $command['command'], '!commandfile' => $command['commandfile'])), 'bootstrap');

          $command_found = TRUE;
          // Dispatch the command(s).
          $return = drush_dispatch($command);

          // prevent a '1' at the end of the output
          if ($return === TRUE) {
            $return = '';
          }

          if (drush_get_context('DRUSH_DEBUG') && !drush_get_context('DRUSH_QUIET')) {
            drush_print_timers();
          }
          drush_log(dt('Peak memory usage was !peak', array('!peak' => drush_format_size(memory_get_peak_usage()))), 'memory');
          break;
        }
      }
    }
    else {
      break;
    }
  }

  if (!$command_found) {
    // If we reach this point, we have not found either a valid or matching command.
    $args = implode(' ', drush_get_arguments());
    if (isset($command) && is_array($command)) {
      foreach ($command['bootstrap_errors'] as $key => $error) {
        drush_set_error($key, $error);
      }
      drush_set_error('DRUSH_COMMAND_NOT_EXECUTABLE', dt("The drush command '!args' could not be executed.", array('!args' => $args)));
    }
    elseif (!empty($args)) {
      drush_set_error('DRUSH_COMMAND_NOT_FOUND', dt("The drush command '!args' could not be found.", array('!args' => $args)));
    }
    // Set errors that ocurred in the bootstrap phases.
    $errors = drush_get_context('DRUSH_BOOTSTRAP_ERRORS', array());
    foreach ($errors as $code => $message) {
      drush_set_error($code, $message);
    }
  }

  drush_bootstrap_finish();

  // After this point the drush_shutdown function will run,
  // exiting with the correct exit code.
  return $return;
}

/**
 * Log the given user in to a bootstrapped Drupal site.
 *
 * @param mixed
 *   Numeric user id or user name.
 *
 * @return boolean
 *   TRUE if user was logged in, otherwise FALSE.
 */
function drush_drupal_login($drush_user) {
  global $user;
  if (drush_drupal_major_version() >= 7) {
    $user = is_numeric($drush_user) ? user_load($drush_user) : user_load_by_name($drush_user);
  }
  else {
    $user = user_load(is_numeric($drush_user) ? array('uid' => $drush_user) : array('name' => $drush_user));
  }

  if (empty($user)) {
    if (is_numeric($drush_user)) {
      $message = dt('Could not login with user ID #!user.', array('!user' => $drush_user));
      if ($drush_user === 0) {
        $message .= ' ' . dt('This is typically caused by importing a MySQL database dump from a faulty tool which re-numbered the anonymous user ID in the users table. See !link for help recovering from this situation.', array('!link' => 'http://drupal.org/node/1029506'));
      }
    }
    else {
      $message = dt('Could not login with user account `!user\'.', array('!user' => $drush_user));
    }
    return drush_set_error('DRUPAL_USER_LOGIN_FAILED', $message);
  }
  else {
    $name = $user->name ? $user->name : variable_get('anonymous', t('Anonymous'));
    drush_log(dt('Successfully logged into Drupal as !name', array('!name' => $name . " (uid=$user->uid)")), 'bootstrap');
  }

  return TRUE;
}
